openbas-api/src/main/java/io/openbas/service/TeamService.java

```diff
@@ -1,21 +1,142 @@
 
 @Service
+@RequiredArgsConstructor
 public class TeamService {
 
-    public Team copyContextualTeam(Team teamToCopy) {
-        Team newTeam = new Team();
-        newTeam.setName(teamToCopy.getName());
-        newTeam.setDescription(teamToCopy.getDescription());
-        newTeam.setTags(CopyObjectListUtils.copy(teamToCopy.getTags(), Tag.class));
-        newTeam.setOrganization(teamToCopy.getOrganization());
-        newTeam.setUsers(CopyObjectListUtils.copy(teamToCopy.getUsers(), User.class));
-        newTeam.setContextual(teamToCopy.getContextual());
-        return newTeam;
+  @PersistenceContext
+  private EntityManager entityManager;
+
+  private final UserRepository userRepository;
+
+  public Team copyContextualTeam(Team teamToCopy) {
+    Team newTeam = new Team();
+    newTeam.setName(teamToCopy.getName());
+    newTeam.setDescription(teamToCopy.getDescription());
+    newTeam.setTags(CopyObjectListUtils.copy(teamToCopy.getTags(), Tag.class));
+    newTeam.setOrganization(teamToCopy.getOrganization());
+    newTeam.setUsers(CopyObjectListUtils.copy(teamToCopy.getUsers(), User.class));
+    newTeam.setContextual(teamToCopy.getContextual());
+    return newTeam;
+  }
+
+  @Tracing(name = "Paginate teams", layer = "service")
+  public Page<TeamOutput> teamPagination(
+      @RequestBody @Valid SearchPaginationInput searchPaginationInput,
+      @NotNull final Specification<Team> teamSpecification) {
+    BiFunction<Specification<Team>, Pageable, Page<TeamOutput>> teamsFunction;
+    OpenBASPrincipal currentUser = currentUser();
+    if (currentUser.isAdmin()) {
+      teamsFunction = (Specification<Team> specification, Pageable pageable) -> this.paginate(
+          teamSpecification.and(specification), pageable
+      );
+    } else {
+      User user = this.userRepository.findById(currentUser.getId()).orElseThrow(ElementNotFoundException::new);
+      List<String> organizationIds = user.getGroups()
+          .stream()
+          .flatMap(group -> group.getOrganizations().stream())
+          .map(Organization::getId)
+          .toList();
+      teamsFunction = (Specification<Team> specification, Pageable pageable) -> this.paginate(
+          teamSpecification.and(teamsAccessibleFromOrganizations(organizationIds)).and(specification), pageable
+      );
+    }
+    return buildPaginationCriteriaBuilder(
+        teamsFunction,
+        searchPaginationInput,
+        Team.class
+    );
+  }
+
+  private Page<TeamOutput> paginate(Specification<Team> specification, Pageable pageable) {
+    CriteriaBuilder cb = this.entityManager.getCriteriaBuilder();
+
+    CriteriaQuery<Tuple> cq = cb.createTupleQuery();
+    Root<Team> teamRoot = cq.from(Team.class);
+    select(cb, cq, teamRoot);
+
+    // -- Specification --
+    if (specification != null) {
+      Predicate predicate = specification.toPredicate(teamRoot, cq, cb);
+      if (predicate != null) {
+        cq.where(predicate);
+      }
+    }
+
```

Comment:
May be we could encapsulate lines 90 to 102 into a separate function to avoid code duplication with the find function. What do you think ?

Reply:
I agree but this is not so easy
The method should return cb, teamRoot and cq.

And I don't want to have a big structure with everything inside